File: app/src/main/java/d/d/meshenger/BackupActivity.java
package d.d.meshenger;

import com.github.isabsent.filepicker.SimpleFilePickerDialog;
import static com.github.isabsent.filepicker.SimpleFilePickerDialog.CompositeMode.FILE_OR_FOLDER_SINGLE_CHOICE;

import android.app.Service;
import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.Environment;
import android.os.IBinder;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import java.io.File;


public class BackupActivity extends MeshengerActivity implements ServiceConnection,
        SimpleFilePickerDialog.InteractionListenerString {
    private static final String SELECT_PATH_REQUEST = "SELECT_PATH_REQUEST";
    private static final int REQUEST_PERMISSION = 0x01;
    private AlertDialog.Builder builder;
    private Button exportButton;
    private Button importButton;
    private ImageButton selectButton;
    private TextView pathEditText;
    private TextView passwordEditText;
    private MainService.MainBinder binder;

    private void showErrorMessage(String title, String message) {
        builder.setTitle(title);
        builder.setMessage(message);
        builder.setPositiveButton(android.R.string.ok, null);
        builder.show();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_backup);

        bindService();
        initViews();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (this.binder != null) {
            unbindService(this);
        }
    }

    private void bindService() {
        // ask MainService to get us the binder object
        Intent serviceIntent = new Intent(this, MainService.class);
        bindService(serviceIntent, this, Service.BIND_AUTO_CREATE);
    }

    @Override
    public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
        this.binder = (MainService.MainBinder) iBinder;
        initViews();
    }

    @Override
    public void onServiceDisconnected(ComponentName componentName) {
        this.binder = null;
    }

    private void initViews() {
        if (this.binder == null) {
            return;
        }

        builder = new AlertDialog.Builder(this);
        importButton = findViewById(R.id.ImportButton);
        exportButton = findViewById(R.id.ExportButton);
        selectButton = findViewById(R.id.SelectButton);
        pathEditText = findViewById(R.id.PathEditText);
        passwordEditText = findViewById(R.id.PasswordEditText);

        importButton.setOnClickListener((View v) -> {
            if (Utils.hasReadPermission(BackupActivity.this)) {
                importDatabase();
            } else {
                Utils.requestReadPermission(BackupActivity.this, REQUEST_PERMISSION);
            }
        });

        exportButton.setOnClickListener((View v) -> {
            if (Utils.hasReadPermission(BackupActivity.this) && Utils.hasWritePermission(BackupActivity.this)) {
                exportDatabase();
            } else {
                Utils.requestWritePermission(BackupActivity.this, REQUEST_PERMISSION);
                Utils.requestReadPermission(BackupActivity.this, REQUEST_PERMISSION);
            }
        });

        selectButton.setOnClickListener((View v) -> {
            if (Utils.hasReadPermission(BackupActivity.this)) {
                final String rootPath = Environment.getExternalStorageDirectory().getAbsolutePath();
                showListItemDialog(getResources().getString(R.string.button_select), rootPath, FILE_OR_FOLDER_SINGLE_CHOICE, SELECT_PATH_REQUEST);
            } else {
                Utils.requestReadPermission(BackupActivity.this, REQUEST_PERMISSION);
            }
        });
    }

    private void exportDatabase() {
        String password = passwordEditText.getText().toString();
        String path = pathEditText.getText().toString();

        if (path == null || path.isEmpty()) {
            showErrorMessage(getResources().getString(R.string.error), getResources().getString(R.string.no_path_selected));
            return;
        }

        if ((new File(path)).isDirectory() || path.endsWith("/")) {
            showErrorMessage(getResources().getString(R.string.error), getResources().getString(R.string.no_file_name));
            return;
        }

        if ((new File(path)).exists()) {
            showErrorMessage(getResources().getString(R.string.error), getResources().getString(R.string.file_already_exists));
            return;
        }

        try {
            Database db = this.binder.getDatabase();
            Database.store(path, db, password);
            Toast.makeText(this, R.string.done, Toast.LENGTH_SHORT).show();
        } catch (Exception e) {
            showErrorMessage(getResources().getString(R.string.error), e.getMessage());
        }
    }

    private void importDatabase() {
        String password = passwordEditText.getText().toString();
        String path = pathEditText.getText().toString();

        if (path == null || path.isEmpty()) {
            showErrorMessage(getResources().getString(R.string.error), getResources().getString(R.string.no_path_selected));
            return;
        }

        if ((new File(path)).isDirectory() || path.endsWith("/")) {
            showErrorMessage(getResources().getString(R.string.error), getResources().getString(R.string.no_file_name));
            return;
        }

        if (!(new File(path)).exists()) {
            showErrorMessage(getResources().getString(R.string.error), getResources().getString(R.string.file_does_not_exist));
            return;
        }

        try {
            Database db = Database.load(path, password);
            this.binder.replaceDatabase(db);
            Toast.makeText(this, R.string.done, Toast.LENGTH_SHORT).show();
        } catch (Exception e) {
            showErrorMessage(getResources().getString(R.string.error), e.toString());
        }
    }

    // path picker
    @Override
    public void showListItemDialog(String title, String folderPath, SimpleFilePickerDialog.CompositeMode mode, String dialogTag) {
        SimpleFilePickerDialog.build(folderPath, mode)
                .title(title)
                .show(this, dialogTag);
    }

    @Override
    public boolean onResult(@NonNull String dialogTag, int which, @NonNull Bundle extras) {
        switch (dialogTag) {
            case SELECT_PATH_REQUEST:
                if (extras.containsKey(SimpleFilePickerDialog.SELECTED_SINGLE_PATH)) {
                    String path = extras.getString(SimpleFilePickerDialog.SELECTED_SINGLE_PATH);
                    if ((new File(path)).isDirectory()) {
                        // append slash
                        if (!path.endsWith("/")) {
                            path += "/";
                        }
                        path += "meshenger_backup.json";
                    }
                    pathEditText.setText(path);
                }
                break;
        }
        return false;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        switch (requestCode) {
            case REQUEST_PERMISSION:
                if (Utils.allGranted(grantResults)) {
                    // permissions granted
                    Toast.makeText(getApplicationContext(), "Permissions granted - please try again.", Toast.LENGTH_SHORT).show();
                } else {
                    showErrorMessage("Permissions Required", "Action cannot be performed.");
                }
                break;
        }
    }

    private void log(String s) {
        Log.d(this, s);
    }
}

File: app/src/main/java/d/d/meshenger/BackupActivity.kt
package d.d.meshenger

import android.app.Activity
import d.d.meshenger.Utils.writeExternalFile
import d.d.meshenger.Utils.readExternalFile
import android.widget.TextView
import android.os.Bundle
import android.content.Intent
import d.d.meshenger.MainService.MainBinder
import android.widget.Toast
import android.content.ComponentName
import android.content.DialogInterface
import android.content.ServiceConnection
import android.net.Uri
import android.os.IBinder
import android.widget.Button
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import java.lang.Exception

class BackupActivity : BaseActivity(), ServiceConnection {
    private var dialog: AlertDialog? = null
    private var binder: MainBinder? = null
    private lateinit var exportButton: Button
    private lateinit var importButton: Button
    private lateinit var passwordEditText: TextView

    private fun showMessage(title: String, message: String) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle(title)
        builder.setMessage(message)
        builder.setPositiveButton(android.R.string.ok, null)
        dialog = builder.show()
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_backup)

        val toolbar = findViewById<Toolbar>(R.id.backup_toolbar)
        toolbar.apply {
            setNavigationOnClickListener {
                finish()
            }
        }
        setSupportActionBar(toolbar)
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowTitleEnabled(false)
        }
        bindService()
        initViews()
    }

    override fun onDestroy() {
        dialog?.dismiss()

        super.onDestroy()

        if (binder != null) {
            unbindService(this)
        }
    }

    private fun bindService() {
        // ask MainService to get us the binder object
        val serviceIntent = Intent(this, MainService::class.java)
        bindService(serviceIntent, this, BIND_AUTO_CREATE)
    }

    override fun onServiceConnected(componentName: ComponentName, iBinder: IBinder) {
        binder = iBinder as MainBinder
        initViews()
    }

    override fun onServiceDisconnected(componentName: ComponentName) {
        binder = null
    }

    private fun initViews() {
        if (binder == null) {
            return
        }

        importButton = findViewById(R.id.ImportButton)
        exportButton = findViewById(R.id.ExportButton)
        passwordEditText = findViewById(R.id.PasswordEditText)
        importButton.setOnClickListener {
            val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
            intent.addCategory(Intent.CATEGORY_OPENABLE)
            intent.type = "application/json"
            importFileLauncher.launch(intent)
        }

        exportButton.setOnClickListener {
            val intent = Intent(Intent.ACTION_CREATE_DOCUMENT)
            intent.addCategory(Intent.CATEGORY_OPENABLE)
            intent.putExtra(Intent.EXTRA_TITLE, "meshenger-backup.json")
            intent.type = "application/json"
            exportFileLauncher.launch(intent)
        }
    }

    private var importFileLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == Activity.RESULT_OK) {
            val intent = result.data ?: return@registerForActivityResult
            val uri = intent.data ?: return@registerForActivityResult
            importDatabase(uri)
        }
    }

    private var exportFileLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
        if (result.resultCode == Activity.RESULT_OK) {
            val intent = result.data ?: return@registerForActivityResult
            val uri: Uri = intent.data ?: return@registerForActivityResult
            exportDatabase(uri)
        }
    }

    private fun exportDatabase(uri: Uri) {
        val password = passwordEditText.text.toString()
        try {
            val database = binder!!.getDatabase()!!
            val data = Database.toData(database, password)

            if (data != null) {
                writeExternalFile(this, uri, data)
                Toast.makeText(this, R.string.done, Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(this, getString(R.string.failed_to_export_database), Toast.LENGTH_SHORT).show()
            }
        } catch (e: Exception) {
            showMessage(getString(R.string.error), e.message ?: "unknown")
        }
    }

    private fun importDatabase(uri: Uri) {
        val binder = this.binder ?: return
        val newDatabase : Database

        try {
            val password = passwordEditText.text.toString()
            val data = readExternalFile(this, uri)
            newDatabase = Database.fromData(data, password)
        } catch (e: Exception) {
            showMessage(getString(R.string.error), e.toString())
            return
        }

        val contactCount = newDatabase.contacts.contactList.size
        val eventCount = newDatabase.events.eventList.size
        val builder = AlertDialog.Builder(this)
        builder.setTitle(R.string.confirm)
        builder.setMessage(String.format(getString(R.string.import_dialog), contactCount, eventCount))
        builder.setCancelable(false) // prevent key shortcut to cancel dialog
        builder.setPositiveButton(R.string.yes) { dialog: DialogInterface, _: Int ->
            binder.getService().mergeDatabase(newDatabase)
            Toast.makeText(this, getString(R.string.done), Toast.LENGTH_SHORT).show()
            dialog.cancel()
        }

        builder.setNegativeButton(R.string.no) { dialog: DialogInterface, _: Int ->
            dialog.cancel()
        }

        // create dialog box
        val alert = builder.create()
        alert.show()
    }
}